File: src/context.ts
import { Command } from 'commander';
import chalk from 'chalk';
import { DaptinClient } from './client';
import { ConfigManager } from './config';
import { createRenderer, Renderer } from './output';

export interface CliContext {
  client: DaptinClient;
  configManager: ConfigManager;
  renderer: Renderer;
}

/**
 * Build a CLI context from a command's options.
 * Creates the DaptinClient, ConfigManager, and Renderer based on CLI flags and config file.
 */
export function getContext(cmd: Command): CliContext {
  const opts = cmd.optsWithGlobals();
  const configManager = new ConfigManager(opts.config);
  const renderer = createRenderer(opts.output || 'table');

  const context = configManager.getCurrentContext();
  const endpoint = opts.endpoint || context?.endpoint;
  const token = opts.token || context?.token;

  if (!endpoint) {
    console.error(chalk.red('Error: No endpoint configured.'));
    console.error('Use "daptin-cli config add <name> <endpoint>" to add a server,');
    console.error('or pass --endpoint <url> on the command line.');
    process.exit(1);
  }

  const client = new DaptinClient(endpoint, token, opts.debug);
  return { client, configManager, renderer };
}

/**
 * Standard error handler for CLI commands.
 * Formats axios errors and other exceptions into user-friendly messages.
 */
export function handleError(error: any): void {
  if (error.response) {
    const status = error.response.status;
    const data = error.response.data;
    const message =
      typeof data === 'string'
        ? data
        : data?.message || data?.errors?.[0]?.title || JSON.stringify(data);
    console.error(chalk.red(`Error ${status}:`), message);
  } else if (error.code === 'ECONNREFUSED') {
    console.error(
      chalk.red('Connection refused.'),
      'Is the Daptin server running at the configured endpoint?'
    );
  } else if (error.code === 'ENOTFOUND') {
    console.error(chalk.red('Host not found.'), 'Check the endpoint URL.');
  } else if (error.code === 'ETIMEDOUT') {
    console.error(chalk.red('Connection timed out.'));
  } else if (error.message) {
    console.error(chalk.red('Error:'), error.message);
  } else {
    console.error(chalk.red('Unknown error occurred.'));
  }
  process.exit(1);
}

/**
 * Extract attributes from JSON:API response objects.
 */
export function extractAttributes(
  objects: Array<{ attributes?: Record<string, any> }>
): Record<string, any>[] {
  return objects.map((obj) => obj.attributes || {});
}

/**
 * Filter columns from a data array.
 */
export function filterColumns(
  data: Record<string, any>[],
  columns: string[]
): Record<string, any>[] {
  if (!columns || columns.length === 0) return data;
  return data.map((row) => {
    const filtered: Record<string, any> = {};
    for (const col of columns) {
      if (col in row) filtered[col] = row[col];
    }
    return filtered;
  });
}

File: cmd/context.go
package cmd

import (
	"fmt"

	"github.com/daptin/daptin-cli/config"
	"github.com/urfave/cli/v2"
)

func contextCommand(appCtx *AppContext) *cli.Command {
	return &cli.Command{
		Name:  "context",
		Usage: "Manage server contexts",
		Subcommands: []*cli.Command{
			{
				Name:      "set",
				Usage:     "Set the active context",
				ArgsUsage: "<name>",
				Action: func(c *cli.Context) error {
					name := c.Args().Get(0)
					if name == "" {
						return fmt.Errorf("context name required")
					}
					if err := appCtx.Config.SetContext(name); err != nil {
						return err
					}
					return appCtx.Config.Save()
				},
			},
			{
				Name:      "add",
				Usage:     "Add a new context",
				ArgsUsage: "<name> <endpoint>",
				Action: func(c *cli.Context) error {
					name := c.Args().Get(0)
					endpoint := c.Args().Get(1)
					if name == "" || endpoint == "" {
						return fmt.Errorf("usage: context add <name> <endpoint>")
					}
					appCtx.Config.UpsertHost(config.HostEndpoint{
						Name:     name,
						Endpoint: endpoint,
					})
					return appCtx.Config.Save()
				},
			},
			{
				Name:  "list",
				Usage: "List all contexts",
				Action: func(c *cli.Context) error {
					if len(appCtx.Config.Hosts) == 0 {
						fmt.Println("No contexts configured")
						return nil
					}
					for _, h := range appCtx.Config.Hosts {
						marker := "  "
						if h.Name == appCtx.Config.CurrentContext {
							marker = "* "
						}
						hasToken := ""
						if h.Token != "" {
							hasToken = " (authenticated)"
						}
						fmt.Printf("%s%s  %s%s\n", marker, h.Name, h.Endpoint, hasToken)
					}
					return nil
				},
			},
		},
	}
}
